### PR TITLE
[FEATURE][locator] Add a goto locator

### PR DESCRIPTION
**Pull request by nirvn**
## Description

Because I felt lame over the weekend showcasing QGIS and not being able to go to a longitude latitude coordinate, we have a brand new goto locator filter. This is co-presented by @Gustry who first had the idea and submitted a PR a while back. 

The goto indicator takes in the following strings:
- a pair of x and y (separated by a comma or a space)
- an openstreetmap/leaflet/openlayer formatted URL
- a google maps URL

The logic of the x y string is as follow:
- if the canvas CRS isn't WGS84:
  - if the coordinate fits within the canvas CRS bounds, add a canvas CRS result
  - if the coordinate fits within WGS84 bounds, add a WGS84 result (which will weight more than the above)
- if the canvas CRS is WGS84:
  - if the coordinate fits within the WGS84 bound, add a WGS84 result

The obligatory show off GIF:
Peek 2020-07-27 11-44

### CODE REVIEW

File: src/app/locator/qgsinbuiltlocatorfilters.cpp [hunk 1]
```diff
@@ -702,3 +708,211 @@ void QgsBookmarkLocatorFilter::triggerResult( const QgsLocatorResult &result )
...
+      bool withinCurrentCrs = currentCrs.bounds().isNull() || currentCrs.bounds().contains( point );
+      bool withinWgs84 = wgs84Crs.bounds().contains( point );
+
+      if ( currentCrs != wgs84Crs )
+      {
+        if ( withinCurrentCrs )
+        {
+          QgsLocatorResult result;
+          result.filter = this;
+          result.displayString = tr( "Go to %1 %2 (%3)" ).arg( QString::number( point.x() ), QString::number( point.y() ), currentCrs.authid() );
```
Comment (nyalldawson) [inline src/app/locator/qgsinbuiltlocatorfilters.cpp:762]:
```suggestion
          result.displayString = tr( "Go to %1 %2 (%3)" ).arg( QString::number( point.x() ), QString::number( point.y() ), currentCrs.userFriendlyIdentifier() );
```

File: src/app/locator/qgsinbuiltlocatorfilters.cpp [hunk 2]
```diff
@@ -702,3 +708,211 @@ void QgsBookmarkLocatorFilter::triggerResult( const QgsLocatorResult &result )
...
+
+        if ( withinWgs84 )
+        {
+          QgsCoordinateTransform transform( wgs84Crs, currentCrs, QgsProject::instance()->transformContext() );
+          QgsPointXY transformedPoint = transform.transform( point );
+          data[QStringLiteral( "point" )] = transformedPoint;
+
+          QgsLocatorResult result;
+          result.filter = this;
+          result.displayString = tr( "Go to %1 %2 (%3)" ).arg( QString::number( point.x() ), QString::number( point.y() ), wgs84Crs.authid() );
```
Comment (nyalldawson) [inline src/app/locator/qgsinbuiltlocatorfilters.cpp:776]:
```suggestion
          result.displayString = tr( "Go to %1 %2 (%3)" ).arg( QString::number( point.x() ), QString::number( point.y() ), wgs84Crs.userFriendlyIdentifier() );
```

File: src/app/locator/qgsinbuiltlocatorfilters.cpp [hunk 3]
```diff
@@ -702,3 +708,211 @@ void QgsBookmarkLocatorFilter::triggerResult( const QgsLocatorResult &result )
...
+          result.userData = data;
+          result.score = 1.0;
+          emit resultFetched( result );
+        }
+      }
+      else
+      {
+        if ( withinWgs84 )
```
Comment (nyalldawson) [inline src/app/locator/qgsinbuiltlocatorfilters.cpp:784]: couldn't this block be combined with the previous `if ( withinWgs84 )` block?

File: src/app/locator/qgsinbuiltlocatorfilters.cpp [hunk 4]
```diff
@@ -702,3 +708,211 @@ void QgsBookmarkLocatorFilter::triggerResult( const QgsLocatorResult &result )
...
+    if ( okX && okY )
+    {
+      QVariantMap data;
+      if ( scale > 0.0 )
+      {
+        data.insert( QStringLiteral( "scale" ), scale );
+      }
+
+      QgsPointXY point( posX, posY );
+      bool withinWgs84 = wgs84Crs.bounds().contains( point );
```
Comment (nyalldawson) [inline src/app/locator/qgsinbuiltlocatorfilters.cpp:880]: won't these always be wgs84 ?

File: src/app/locator/qgsinbuiltlocatorfilters.cpp [hunk 5]
```diff
@@ -702,3 +708,211 @@ void QgsBookmarkLocatorFilter::triggerResult( const QgsLocatorResult &result )
...
+      else
+      {
+        data.insert( QStringLiteral( "point" ), point );
+      }
+
+      QgsLocatorResult result;
+      result.filter = this;
+      result.displayString = tr( "Go to %1 %2 %3(%4)" ).arg( point.x() )
+                             .arg( point.y() )
+                             .arg( scale > 0.0 ? QStringLiteral( "at scale 1:%1 " ).arg( scale ) : QString() )
```
Comment (nyalldawson) [inline src/app/locator/qgsinbuiltlocatorfilters.cpp:896]:
```suggestion
                             .arg( scale > 0.0 ? tr( "at scale 1:%1 " ).arg( scale ) : QString() )
```

File: src/app/locator/qgsinbuiltlocatorfilters.cpp [hunk 6]
```diff
@@ -702,3 +708,211 @@ void QgsBookmarkLocatorFilter::triggerResult( const QgsLocatorResult &result )
...
+                             .arg( wgs84Crs.authid() );
```
Comment (nyalldawson) [inline src/app/locator/qgsinbuiltlocatorfilters.cpp:897]:
```suggestion
                             .arg( wgs84Crs.userFriendlyIdentifier() );
```

File: tests/src/app/testqgsapplocatorfilters.cpp [hunk 7]
```diff
@@ -272,5 +273,32 @@ QList<QgsLocatorResult> TestQgsAppLocatorFilters::gatherResults( QgsLocatorFilte
...
+void TestQgsAppLocatorFilters::testGoto()
+{
+  QgsGotoLocatorFilter filter;
+
+  // goto X,Y
+  QList< QgsLocatorResult > results = gatherResults( &filter, QStringLiteral( "4 5" ), QgsLocatorContext() );
+  QCOMPARE( results.count(), 2 );
+  QCOMPARE( results.at( 0 ).displayString, QStringLiteral( "Go to 4 5 ()" ) );
+  QCOMPARE( results.at( 0 ).userData.toMap()[QStringLiteral( "point" )].value<QgsPointXY>(), QgsPointXY( 4, 5 ) );
+  QCOMPARE( results.at( 1 ).displayString, QStringLiteral( "Go to 4 5 (EPSG:4326)" ) );
```
Comment (nyalldawson) [inline tests/src/app/testqgsapplocatorfilters.cpp:285]: For geographic CRS, would be nice to use a degrees suffix here

**Review by nyalldawson — COMMENTED** [review]
Yay, happy to see this resurrected!!

File: src/app/locator/qgsinbuiltlocatorfilters.cpp (hunk 4 above)
Comment (nirvn) [reply src/app/locator/qgsinbuiltlocatorfilters.cpp:880]: Yes, except when someone passes on a flawed URL with erroneous lon and/or lat, which then leads to a transformation error. I'd rather deal with those via a bounds check than an error catch.

**Comment by Gustry** [comment]
Thanks @nirvn for taking back this old PR !

**Comment by nirvn** [comment]
@Gustry , thanks for the initial PR! :)

**Comment by nirvn** [comment]
@nyalldawson , comments addressed, OK to merge?

File: src/app/locator/qgsinbuiltlocatorfilters.cpp [hunk 8]
```diff
@@ -702,3 +708,205 @@ void QgsBookmarkLocatorFilter::triggerResult( const QgsLocatorResult &result )
...
+            }
+            posX = params.at( 2 ).toDouble( &okX );
+            posY = params.at( 1 ).toDouble( &okY );
+          }
+          break;
+        }
+      }
+    }
+
+    if ( !okX && !okY && string.indexOf( QStringLiteral( "google.com/maps/" ) ) != -1 )
```
Comment (uclaros) [inline src/app/locator/qgsinbuiltlocatorfilters.cpp:847]: It would be nice to not limit this to `.com` but also handle international TLDs

Comment (nirvn) [reply src/app/locator/qgsinbuiltlocatorfilters.cpp:847]: @uclaros , URL example?

Comment (nirvn) [reply src/app/locator/qgsinbuiltlocatorfilters.cpp:847]: No worries, got i, will adjust accordingly. Good catch.

File: src/app/locator/qgsinbuiltlocatorfilters.cpp [hunk 9]
```diff
@@ -702,3 +708,201 @@ void QgsBookmarkLocatorFilter::triggerResult( const QgsLocatorResult &result )
...
+
+  // Coordinates such as 106.8468,-6.3804
+  QRegularExpression separatorRx( QStringLiteral( "^([0-9\\-\\.]*)[\\s\\,]*([0-9\\-\\.]*)$" ) );
+  QRegularExpressionMatch match = separatorRx.match( string.trimmed() );
+  if ( match.hasMatch() )
+  {
+    bool okX = false;
+    bool okY = false;
+    double posX = match.captured( 1 ).toDouble( &okX );
+    double posY = match.captured( 2 ).toDouble( &okY );
```
Comment (uclaros) [inline src/app/locator/qgsinbuiltlocatorfilters.cpp:741]: Will this handle different locales correctly or maybe we need something like `QLocale::toDouble`?

Comment (nirvn) [reply src/app/locator/qgsinbuiltlocatorfilters.cpp:741]: Locale stuff will come in a follow up PR.

**Comment by timlinux** [comment]
> a pair of x and y (separated by a comma or a space)

Isn't this case already covered by typing the coordinates into the coordinates widget in the status bar?

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
@timlinux 
> Isn't this case already covered by typing the coordinates into the coordinates widget in the status bar?

That only works in the current canvas CRS. So if you have a map in a projected CRS and want to zoom to a particular lat/long, you're completely out of luck :laughing:

**Comment by nirvn** [comment]
@timlinux , as Nyall mentioned, the status bar coordinates widget only support pasting Xs and Ys in the project/canvas CRS. Expanding on the above-mentioned shame, I had to temporarily switch my project's UTM CRS to WGS84 merely to paste a lon lat. Unacceptable :+1: 

The locator's UI (i.e. the list of available locators showing up upon clicking on text entry) makes for better discoverability of the feature too.

**Comment by nyalldawson** [comment]
> The locator's UI (i.e. the list of available locators showing up upon clicking on text entry) makes for better discoverability of the feature too.

Also, it's much more convenient because you can trigger the search with Ctrl + k, vs the status bar widget which requires a mouse click to focus.

**Comment by nirvn** [comment]
@nyalldawson , true that.

**Pull request merged**